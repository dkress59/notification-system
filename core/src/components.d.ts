import { NotificationType } from "./types";
export namespace Components {
	interface BannerArea {
		/**
		  * Takes the `<banner-notification />`-component's props in camelCase as an argument
		 */
		"spawnBanner": ({ autoHide, autoHideAfterMs, content, headline, type, }: { autoHide?: boolean; autoHideAfterMs?: number; content: string; headline?: string; type?: NotificationType; }) => void;
	}
	interface BannerNotification {
		/**
		  * Whether to automatically hide the banner, or not. If false (or undefined), a dismiss-button will be rendered.
		 */
		"autoHide": boolean;
		/**
		  * The time in milliseconds after which the banner shall be hidden (requires the auto-hide attribute to be set to "true").
		 */
		"autoHideAfterMs": number;
		/**
		  * Entirely dismisses the banner from the DOM
		 */
		"dismiss": () => void;
		/**
		  * If provided, the banner will be rendered with a headline which is styled slightly more prominent than the body text.
		 */
		"headline": string;
		/**
		  * The notification-type of the banner (success | info | warning | error).
		 */
		"type": NotificationType;
	}
	interface ModalNotification {
		/**
		  * If set to 'false' the 'Confirm'-button will be disabled.
		 */
		"condition": boolean | undefined;
		/**
		  * Entirely dismisses the modal from the DOM
		 */
		"dismiss": () => void;
		/**
		  * If provided, the modal will be rendered with a headline which is styled slightly more prominent than the body text.
		 */
		"headline": string;
		/**
		  * Label for the 'Confirm'-button
		 */
		"labelConfirm": string;
		/**
		  * Label for the 'Decline'-button
		 */
		"labelDecline": string;
		/**
		  * Whether to show the 'Confirm'-button.
		 */
		"showConfirm": boolean;
		/**
		  * Whether to show the 'Decline'-button.
		 */
		"showDecline": boolean;
		/**
		  * The notification-type of the modal (success | info | warning | error).
		 */
		"type": NotificationType;
	}
	interface NotificationArea {
		/**
		  * Takes the `<modal-notification />`-component's props in camelCase as an argument
		 */
		"spawnModal": ({ condition, content, headline, showConfirm, showDecline, labelConfirm, labelDecline, type, }: { condition?: boolean; content: string; headline?: string; showConfirm?: boolean; showDecline?: boolean; labelConfirm?: string; labelDecline?: string; type?: NotificationType; }) => void;
		/**
		  * Takes the `<toast-notification />`-component's props in camelCase as an argument
		 */
		"spawnToast": ({ autoHide, autoHideAfterMs, content, headline, type, }: { autoHide?: boolean; autoHideAfterMs?: number; content: string; headline?: string; type?: NotificationType; }) => void;
	}
	interface ToastNotification {
		/**
		  * Whether to automatically hide the toast, or not. If false (or undefined), a dismiss-button will be rendered.
		 */
		"autoHide": boolean;
		/**
		  * The time in milliseconds after which the toast shall be hidden (requires the auto-hide attribute to be set to "true").
		 */
		"autoHideAfterMs": number;
		/**
		  * Entirely dismisses the toast from the DOM
		 */
		"dismiss": () => void;
		/**
		  * If provided, the toast will be rendered with a headline which is styled slightly more prominent than the body text.
		 */
		"headline": string;
		/**
		  * The notification-type of the toast (success | info | warning | error).
		 */
		"type": NotificationType;
	}
}
export namespace Elements {
	interface HTMLBannerAreaElement extends Components.BannerArea, HTMLElement {
	}
	var HTMLBannerAreaElement: {
		prototype: HTMLBannerAreaElement;
		new(): HTMLBannerAreaElement;
	};
	interface HTMLBannerNotificationElement extends Components.BannerNotification, HTMLElement {
	}
	var HTMLBannerNotificationElement: {
		prototype: HTMLBannerNotificationElement;
		new(): HTMLBannerNotificationElement;
	};
	interface HTMLModalNotificationElement extends Components.ModalNotification, HTMLElement {
	}
	var HTMLModalNotificationElement: {
		prototype: HTMLModalNotificationElement;
		new(): HTMLModalNotificationElement;
	};
	interface HTMLNotificationAreaElement extends Components.NotificationArea, HTMLElement {
	}
	var HTMLNotificationAreaElement: {
		prototype: HTMLNotificationAreaElement;
		new(): HTMLNotificationAreaElement;
	};
	interface HTMLToastNotificationElement extends Components.ToastNotification, HTMLElement {
	}
	var HTMLToastNotificationElement: {
		prototype: HTMLToastNotificationElement;
		new(): HTMLToastNotificationElement;
	};
	interface HTMLElementTagNameMap {
		"banner-area": HTMLBannerAreaElement;
		"banner-notification": HTMLBannerNotificationElement;
		"modal-notification": HTMLModalNotificationElement;
		"notification-area": HTMLNotificationAreaElement;
		"toast-notification": HTMLToastNotificationElement;
	}
}
declare namespace LocalJSX {
	interface BannerArea {
	}
	interface BannerNotification {
		/**
		  * Whether to automatically hide the banner, or not. If false (or undefined), a dismiss-button will be rendered.
		 */
		"autoHide"?: boolean;
		/**
		  * The time in milliseconds after which the banner shall be hidden (requires the auto-hide attribute to be set to "true").
		 */
		"autoHideAfterMs"?: number;
		/**
		  * If provided, the banner will be rendered with a headline which is styled slightly more prominent than the body text.
		 */
		"headline"?: string;
		/**
		  * Fires after the elements has transitioned out.
		 */
		"onBannerDismissed"?: (event: CustomEvent<HTMLElement>) => void;
		/**
		  * The notification-type of the banner (success | info | warning | error).
		 */
		"type"?: NotificationType;
	}
	interface ModalNotification {
		/**
		  * If set to 'false' the 'Confirm'-button will be disabled.
		 */
		"condition"?: boolean | undefined;
		/**
		  * If provided, the modal will be rendered with a headline which is styled slightly more prominent than the body text.
		 */
		"headline"?: string;
		/**
		  * Label for the 'Confirm'-button
		 */
		"labelConfirm"?: string;
		/**
		  * Label for the 'Decline'-button
		 */
		"labelDecline"?: string;
		/**
		  * Fires when the 'Confirm'-button is pressed.
		 */
		"onConfirmTriggered"?: (event: CustomEvent<HTMLElement>) => void;
		/**
		  * Fires when the 'Decline'-button is pressed.
		 */
		"onDeclineTriggered"?: (event: CustomEvent<HTMLElement>) => void;
		/**
		  * Fires after the elements has transitioned out.
		 */
		"onModalDismissed"?: (event: CustomEvent<HTMLElement>) => void;
		/**
		  * Whether to show the 'Confirm'-button.
		 */
		"showConfirm"?: boolean;
		/**
		  * Whether to show the 'Decline'-button.
		 */
		"showDecline"?: boolean;
		/**
		  * The notification-type of the modal (success | info | warning | error).
		 */
		"type"?: NotificationType;
	}
	interface NotificationArea {
	}
	interface ToastNotification {
		/**
		  * Whether to automatically hide the toast, or not. If false (or undefined), a dismiss-button will be rendered.
		 */
		"autoHide"?: boolean;
		/**
		  * The time in milliseconds after which the toast shall be hidden (requires the auto-hide attribute to be set to "true").
		 */
		"autoHideAfterMs"?: number;
		/**
		  * If provided, the toast will be rendered with a headline which is styled slightly more prominent than the body text.
		 */
		"headline"?: string;
		/**
		  * Fires after the elements has transitioned out.
		 */
		"onToastDismissed"?: (event: CustomEvent<HTMLElement>) => void;
		/**
		  * The notification-type of the toast (success | info | warning | error).
		 */
		"type"?: NotificationType;
	}
	interface IntrinsicElements {
		"banner-area": BannerArea;
		"banner-notification": BannerNotification;
		"modal-notification": ModalNotification;
		"notification-area": NotificationArea;
		"toast-notification": ToastNotification;
	}
}
