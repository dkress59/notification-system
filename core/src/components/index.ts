export { BannerArea } from './banner-area/banner-area'
export { BannerNotification } from './banner-notification/banner-notification'
export { ModalNotification } from './modal-notification/modal-notification'
export { NotificationArea } from './notification-area/notification-area'
export { ToastNotification } from './toast-notification/toast-notification'
