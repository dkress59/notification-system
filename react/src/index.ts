export * from './components/banner-area'
export * from './components/banner-notification'
export * from './components/modal-notification'
export * from './components/notification-area'
export * from './components/toast-notification'
export * from './types'
