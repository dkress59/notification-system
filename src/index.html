<!DOCTYPE html>
<html dir="ltr" lang="en">

<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=5.0" />
  <title>Notification Web Components</title>

  <script type="module" src="/build/notification-system.esm.js"></script>
  <script nomodule src="/build/notification-system.js"></script>

  <style>
    * {
      box-sizing: border-box;
    }

    body {
      font-family: sans-serif;
    }

    #notification-container {
      position: fixed;
      z-index: 50;
      top: 0;
      right: 0;
      bottom: 0;
      left: 0;
      pointer-events: none;
      display: flex;
      flex-direction: column;
      align-items: flex-end;
      overflow-y: auto;
    }

    /* #notification-container > * {
        right: 0;
      } */
  </style>
  <script>
    document.addEventListener('DOMContentLoaded', () => {
      document.querySelectorAll('notify-toast').forEach(toast => {
        toast.addEventListener('toastDismissed', (shadowRoot) => console.debug(
          'toast dismissed!',
          { shadowRoot }
        ))
      })
    })
  </script>
</head>

<body>
  <div id="notification-container">
    <!-- <my-component first="Stencil" last="'Don't call me a framework' JS"></my-component>
      <my-component first="more Stencil" middle="please" last="'Don't call me a framework' JS"></my-component> -->
    <notify-toast type="error" headline="Lorem Ipsum">Lorem ipsum dolor sit amet consectetur adipisicing elit, facilis
      necessitatibus sunt quisquam officia quidem.</notify-toast>
    <notify-toast>Mock Content</notify-toast>
    <notify-toast auto-hide>Mock Content</notify-toast>
    <notify-toast type="info" is-hidden>Hidden Content</notify-toast>
    <notify-toast>Mock Content</notify-toast>
  </div>
</body>

</html>