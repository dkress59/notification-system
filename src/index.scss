@import 'css/base.scss';
// @import 'css/util.scss';
