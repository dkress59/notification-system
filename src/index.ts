export { Components, JSX } from './components'
export { BannerArea } from './components/banner-area/banner-area'
export { BannerNotification } from './components/banner-notification/banner-notification'
export { ModalNotification } from './components/modal-notification/modal-notification'
export { NotificationArea } from './components/notification-area/notification-area'
export { ToastNotification } from './components/toast-notification/toast-notification'
