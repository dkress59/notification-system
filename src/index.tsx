import './index.scss'

import React from 'react'
import ReactDOM from 'react-dom'

import { App } from './app'
import { NotificationProvider } from './hooks'

ReactDOM.render(
	<React.StrictMode>
		<NotificationProvider>
			<App />
		</NotificationProvider>
	</React.StrictMode>,
	document.getElementById('app'),
)
