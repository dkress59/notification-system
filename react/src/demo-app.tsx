import React, { useEffect, useRef, useState } from 'react'
import { createRoot } from 'react-dom/client'

import { NotificationType } from ':core/types'

import { BannerArea } from './components/banner-area'
import { BannerNotification } from './components/banner-notification'
import { ModalNotification } from './components/modal-notification'
import { NotificationArea } from './components/notification-area'
import { ToastNotification } from './components/toast-notification'

function DemoApp() {
	const [forceReRender, setForceReRender] = useState(false)
	const [bannerChild2, setBannerChild2] = useState(false)
	const notificationAreaRef = useRef<null | HTMLNotificationAreaElement>(null)
	const bannerRef = useRef<null | HTMLBannerNotificationElement>(null)
	const modalRef = useRef<null | HTMLModalNotificationElement>(null)

	useEffect(() => {
		notificationAreaRef.current?.spawnModal({
			content: 'yoyoyo',
			headline: 'Modal spawned',
			type: NotificationType.WARNING,
		})
		const to1 = setTimeout(() => setBannerChild2(true), 2000)
		const to2 = setTimeout(() => bannerRef.current?.dismiss(), 5000)
		const to3 = setTimeout(() => setForceReRender(true), 8000)
		return () => {
			clearTimeout(to1)
			clearTimeout(to2)
			clearTimeout(to3)
		}
	}, [])

	return (
		<>
			<BannerArea>
				<BannerNotification
					headline="some headline"
					ref={bannerRef}
					type={NotificationType.ERROR}
				>
					demo content
				</BannerNotification>
				{bannerChild2 && (
					<BannerNotification
						headline="some headline 2"
						type={NotificationType.INFO}
						onDismiss={() => alert('banner hidden!')}
					>
						demo content
					</BannerNotification>
				)}
			</BannerArea>
			{forceReRender && <>Re-Rendered</>}
			<NotificationArea ref={notificationAreaRef}>
				<ToastNotification
					headline="Info Toast"
					type={NotificationType.INFO}
				>
					Toast Content
				</ToastNotification>
				<ToastNotification
					headline="Toast title"
					autoHide={true}
					autoHideAfterMs={1000}
					onDismiss={() => alert('toast auto-hidden!')}
				/>
				<ModalNotification
					condition={true}
					labelConfirm="Accept"
					onConfirm={() => alert('modal confirmed!')}
					ref={modalRef}
				>
					Modal content
				</ModalNotification>
			</NotificationArea>
		</>
	)
}

const container = document.getElementById('app')!

const root = createRoot(container)
root.render(<DemoApp />)
