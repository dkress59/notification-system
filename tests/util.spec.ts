import { fireEvent } from '@testing-library/dom'

import { NotificationType } from '../src'
import {
	getClosingButtonElement,
	getHeadlineElement,
	getIconElement,
	getStyleElement,
} from '../src/utils'

describe('util.ts', () => {
	describe('getIconElement', () => {
		it('type=error matches snapshot', () => {
			expect(getIconElement(NotificationType.ERROR)).toMatchSnapshot()
		})
		it('type=info matches snapshot', () => {
			expect(getIconElement(NotificationType.INFO)).toMatchSnapshot()
		})
		it('type=success matches snapshot', () => {
			expect(getIconElement(NotificationType.SUCCESS)).toMatchSnapshot()
		})
		it('type=warning matches snapshot', () => {
			expect(getIconElement(NotificationType.WARNING)).toMatchSnapshot()
		})
	})
	describe('getHeadlineElement', () => {
		it('headline matches snapshot', () => {
			expect(getHeadlineElement('mock_headline')).toMatchSnapshot()
		})
		it('undefined returns undefined', () => {
			expect(getHeadlineElement(undefined)).toBeUndefined()
		})
	})
	describe('getClosingButtonElement', () => {
		it('matches snapshot', () => {
			expect(getClosingButtonElement(jest.fn())).toMatchSnapshot()
		})
		it('fires when clicked', () => {
			const mockCallback = jest.fn()
			const closingButton = getClosingButtonElement(mockCallback)
			fireEvent.mouseDown(closingButton)
			expect(mockCallback).toHaveBeenCalled()
		})
		it('fires when enter pressed', () => {
			const mockCallback = jest.fn()
			const closingButton = getClosingButtonElement(mockCallback)
			fireEvent.keyDown(closingButton, { key: 'Enter' })
			expect(mockCallback).toHaveBeenCalled()
		})
	})
	describe('getStyleElement', () => {
		it('matches snapshot', () => {
			expect(
				getStyleElement('mock-selector { mock-property: mock-value; }'),
			).toMatchSnapshot()
		})
	})
})
