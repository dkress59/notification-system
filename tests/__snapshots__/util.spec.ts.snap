// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`util.ts getClosingButtonElement matches snapshot 1`] = `
<span
  aria-role="button"
  class="dismiss"
  data-testid="btn-dismiss"
  tabindex="0"
>
  ✕
</span>
`;

exports[`util.ts getHeadlineElement headline matches snapshot 1`] = `
<h4
  data-testid="headline"
>
  mock_headline
</h4>
`;

exports[`util.ts getIconElement type=error matches snapshot 1`] = `
<span
  class="icon"
  data-testid="icon-type"
>
  <span
    class="error"
  >
    ╲
  </span>
</span>
`;

exports[`util.ts getIconElement type=info matches snapshot 1`] = `
<span
  class="icon"
  data-testid="icon-type"
>
  i
</span>
`;

exports[`util.ts getIconElement type=success matches snapshot 1`] = `
<span
  class="icon"
  data-testid="icon-type"
>
  ✓
</span>
`;

exports[`util.ts getIconElement type=warning matches snapshot 1`] = `
<span
  class="icon"
  data-testid="icon-type"
>
  !
</span>
`;

exports[`util.ts getStyleElement matches snapshot 1`] = `
<style>
  mock-selector { mock-property: mock-value; }
</style>
`;
